feature:support random delayed compression

OpenZYK opened the pull request:
Fixes #4012 

At 0 o 'clock every day, a large number of machines begin to compress the log files of the previous day, resulting in the sharp rise of the disk IO of the container, which brings risk to the stability of the cloud disk. Now, the compression random delay x seconds is introduced to reduce the concentration of the compression trigger time and reduce the pressure on the disk.

 ```
         <RollingFile name="RollingFile" 
                     fileName="logs/app.log" 
                     filePattern="logs/app-%d{MM-dd-yyyy}-%i.log.gz">
            <PatternLayout>
                <Pattern>%d %p %c{1.} [%t] %m%n</Pattern>
            </PatternLayout>
            <Policies>
                <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
                <SizeBasedTriggeringPolicy size="10 MB"/>
            </Policies>
            <!-- deferred compression 1s~30s -->
            <DefaultRolloverStrategy max="10" delayedCompressionSeconds="30"/>
        </RollingFile>
```

Commit:
feature:support random delay compression

Commit:
rename DeferredCompressionAction

Mrkell76 commented:
@

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+/**
+ * Wrapper action that schedules compression for defferred execution.
+ * This action wraps another action and schedules it to be executed
+ * after a random defer within a specified time window.
```
```suggestion
 * Decorates an {@link Action} to execute it after a certain amount of delay.
```

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+    private boolean interrupted = false;
+
+    /**
+     * Creates a new DelayedCompressionAction.
```
```suggestion
     * Creates a new instance.
```

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Creates a new DelayedCompressionAction.
+     *
+     * @param originalAction the action to be executed with defer
```
```suggestion
     * @param originalAction the action to be executed with delay
```

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+ * This action wraps another action and schedules it to be executed
+ * after a random defer within a specified time window.
+ */
+public class DeferredCompressionAction implements Action {
```
I think your earlier attempt, `DelayedCompressionAction`, fits the bill better. AFAICT, `deferred` waits for a condition, which does not necessarily needs to be a time, whereas `delay` makes it clear the involvement of time.

Once you implement this change, please update all the usages of `[Dd]efer` in this PR.

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+     *
+     * @return true if the action was successfully executed
+     * @throws IOException if an error occurs during execution
+     */
```
```suggestion
```

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+     */
+    @Override
+    public boolean execute() throws IOException {
+        if (interrupted) {
```
What about `complete == true`?

vy commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/DeferredCompressionAction.java:
```diff
@@ -0,0 +1,200 @@
+     * @param originalAction the action to be executed with defer
+     * @param maxDeferSeconds the maximum defer in seconds (0 means immediate execution)
+     */
+    public DeferredCompressionAction(Action originalAction, int maxDeferSeconds) {
```
_Question:_ Shouldn't we be receiving a range instead of an upper bound?

vy reviewed (changes requested):
@OpenZYK, I think the suggested feature sounds legitimate. I've skimmed through the changes, and dropped some remarks. Would you also mind providing a test and a changelog entry, please?

vy commented:
@OpenZYK, [TimeBasedTriggeringPolicy](https://logging.apache.org/log4j/2.x/manual/appenders/rolling-file.html#TimeBasedTriggeringPolicy) receives a `maxRandomDelay` configuration attribute — doesn't that precisely address the problem you're describing?

OpenZYK commented:
[TimeBasedTriggeringPolicy](https://logging.apache.org/log4j/2.x/manual/appenders/rolling-file.html#TimeBasedTriggeringPolicy) receives a maxRandomDelay configuration attribute —  This is a parameter for scrolling files, which may cause the logs of the first day to be written to the files of the next day. Splitting files itself is not very IO expensive, but what I'm addressing here is the impact of compression on disk IO. I will add some tests based on your suggestion

vy commented:
Can't we change the default behavior of the `TimeBasedTriggeringPolicy.maxRandomDelay` configuration attribute such that it will not affect the cut off time? That is, the roll over will happen precisely at the right time, but only the compression will honor the delay? @ppkarwasz, WDYT?

ppkarwasz commented:
@vy,

I think delaying **when a rollover is triggered** and delaying **when the asynchronous rollover actions execute** should have separate configuration knobs.

The execution of asynchronous actions lives in `RollingFileManager`, so tying that behavior to an attribute on a specific `TriggeringPolicy` (or rollover strategy) feels like a layering violation.

For this PR specifically, I believe there’s a simpler and cleaner approach: instead of wrapping actions and sleeping, we could replace the `ExecutorService` in `RollingFileManager` with a `ScheduledExecutorService` and apply the delay at scheduling time.

OpenZYK commented:
@ppkarwasz Your suggestion is great! I will follow this idea and add some test cases

Commit:
The ScheduledExecutorService is used to realize the delayed compression function, replacing the sleep in the original Action

Commit:
evert imports to avoid unnecessary changes

Commit:
add comment

OpenZYK committed:
Merge branch '2.x' into feature/2.x/delayed-compress

ppkarwasz commented:
@OpenZYK,

To be clear what I suggested was to start the `AsyncAction` from `RollingFileManager` with a delay, not to start delayed actions from within `DelayedCompressAction`: the rollover process has many inter-dependent actions and it is much simpler to execute them sequentially. If you execute them in parallel, you will encounter many subtle bugs.

Commit:
Move the action submission to the RollingFileManager

vy commented:
@OpenZYK, we were expecting you to only touch `RollingFileManager`, and there schedule the action with an arbitrary delay. But we still see `DelayedCompressionAction`. Are we still on the same page?

Commit:
feat(rolling): implement delayed compression via minDelaySeconds/maxDelaySeconds

Replace the single delayedCompressionSeconds parameter (and the
sleep-based DelayedCompressionAction) with a proper scheduler-driven
approach, as suggested in the PR #4011 review.

Changes:
- Remove DelayedCompressionAction and Schedulable: delay logic no
  longer lives inside the Action; sleeping inside an action is an
  anti-pattern that blocks the executor thread.
- Add getMinAsyncDelay()/getMaxAsyncDelay() default methods to
  RolloverDescription so callers can read the configured range without
  casting.
- Extend RolloverDescriptionImpl with a new constructor that accepts
  minAsyncDelay and maxAsyncDelay (in seconds). The constructor
  validates that max >= min and throws IllegalArgumentException
  otherwise.
- Refactor DefaultRolloverStrategy: replace delayedCompressionSeconds
  with minDelaySeconds/maxDelaySeconds builder attributes; rollover()
  now returns a RolloverDescriptionImpl carrying the delay range
  instead of wrapping the action in a DelayedCompressionAction.
- Refactor RollingFileManager: replace the ad-hoc EmptyQueue /
  Schedulable executor with a standard ScheduledThreadPoolExecutor(1).
  Add scheduleAsyncAction(action, min, max) which picks a uniformly
  random delay in [min, max] seconds and submits the action via
  ScheduledExecutorService.schedule(), keeping the Action stateless.
- Add DelayedCompressionTest (10 test cases) covering: descriptor
  defaults, delay-range storage, invalid-range rejection, strategy
  builder defaults and validation, descriptor propagation through
  rollover(), immediate execution (delay=0), scheduled execution with
  a fixed delay, and random-delay range verification.
- Add changelog entry for GH-4011.

Fixes #4011

OpenZYK commented:
Hi @vy @ppkarwasz, removed `DelayedCompressionAction` entirely, replaced executor with `ScheduledThreadPoolExecutor`, split delay into `minDelaySeconds`/`maxDelaySeconds`, added 10 test cases and changelog entry. Would you mind taking another look?

OpenZYK commented:
Hi @vy @ppkarwasz, we noticed that #4071 has been merged into 2.26.0, which adds compression delay support at the `Action` level (`GzCompressAction`/`ZipCompressAction`).

Our approach in this PR is different: we apply the delay at the `RollingFileManager` scheduling layer using `ScheduledThreadPoolExecutor`, which aligns with @ppkarwasz's earlier suggestion to avoid executing actions in parallel and prevent subtle bugs.

Could you let us know whether this PR is still worth pursuing, or if the feature is considered covered by #4071? Happy to close it if it's no longer needed.

vy commented:
> Hi @vy @ppkarwasz, we noticed that #4071 has been merged into 2.26.0,

No, **fortunately** it doesn't! :sweat_smile: It was merged to `2.x` and planned for `2.27.0`.

> which adds compression delay support at the `Action` level (`GzCompressAction`/`ZipCompressAction`).
> 
> Our approach in this PR is different: we apply the delay at the `RollingFileManager` scheduling layer using `ScheduledThreadPoolExecutor`, which aligns with @ppkarwasz's earlier suggestion to avoid executing actions in parallel and prevent subtle bugs.
> 
> Could you let us know whether this PR is still worth pursuing, or if the feature is considered covered by #4071? Happy to close it if it's no longer needed.

@OpenZYK, it is super awesome that you still follow this feature request. :star_struck: I will take care of submitting a new PR steering the incorrectly implemented #4071 in the direction @ppkarwasz suggested, and ask for your review.
